app/models/user.rb

```diff
@@ -5,6 +5,8 @@ class User < ApplicationRecord
   attr_accessible \
     :email, # str: balidated with Devise
     :password, # str: hashed with bcrypt
+    :otp_required_for_login, # boolean
+    :otp_secret, # str
```

Comment:
This might be a problem: if `:otp_secret` is passed in from a form, then it will overwrite what's already there. But that's something we'd never want anyone to overwrite. If we weren't using `attr_accessible` than strong parameters would help us here. But we're not so we need to make sure otherwise we're okay.

Is there a reason why you chose to make otp_secret accessible here?

Reply:
This throws an error if it's not accessible:

```ruby
User.create!(attrs)
```

If it's not accessible you have to do something like this after the user is created, and probably wrap it all in a transaction:

```ruby
      user.otp_required_for_login = true
      user.otp_secret = User.generate_otp_secret
      user.save!
```

Which is fine and I suppose worth it if it makes things more secure. Just less clean.

Reply:
Oh, it absolutely is less clean. That's why Rails moved to strong parameters. Until we switch though, we have to work with `attr_accessible`